# the R script below is adapted from Daniel Beiting and modified by me

# load packages----
library(rhdf5) #provides functions for handling hdf5 file formats (kallisto outputs bootstraps in this format)
library(tidyverse) # provides access to Hadley Wickham's collection of R packages for data science, which we will use throughout the course
library(tximport) # package for getting Kallisto results into R
library(ensembldb) #helps deal with ensembl
library(EnsDb.Mmusculus.v79) #replace with your organism-specific database package
library(beepr) #just for fun
library(datapasta) #great for copy/paste data into the R environment

# ==============================================================================

# read in your study design ----
#there are LOTS of ways to read data into R, but the readr package (from tidyverse) is one of the simplest
targets <- read_tsv("studydesign.txt")

# you can easily create file paths to the abundance files generated by Kallisto using the 'file.path' function
path <- file.path(targets$sample, "abundance.tsv") # set file paths to your mapped data

# now check to make sure this path is correct by seeing if the files exist
all(file.exists(path)) 

# ==============================================================================

# get annotations using organism-specific package ----
Tx <- transcripts(EnsDb.Mmusculus.v79, columns=c("tx_id", "gene_name"))
Tx <- as_tibble(Tx)

#need to change first column name to 'target_id'
Tx <- dplyr::rename(Tx, target_id = tx_id)

#transcrip ID needs to be the first column in the dataframe
Tx <- dplyr::select(Tx, "target_id", "gene_name")

write.csv(Tx, "../../tx.csv")

# ==============================================================================
# OPTIONAL: get annotations using BiomaRt----
library(biomaRt) # an alternative for annotation

listMarts() #default host is ensembl.org, and most current release of mammalian genomes
#listMarts(host="parasite.wormbase.org") #access to parasite worm genomes
#listMarts(host="protists.ensembl.org") #access to protozoan genomes

#choose the 'mart' you want to work with
myMart <- useMart(biomart="ENSEMBL_MART_ENSEMBL")
#take a look at all available datasets within the selected mart
available.datasets <- listDatasets(myMart)
#now grab the ensembl annotations for dog
dog.anno <- useMart(biomart="ENSEMBL_MART_ENSEMBL", dataset = "clfamiliaris_gene_ensembl")
dog.attributes <- listAttributes(dog.anno)
Tx.dog <- getBM(attributes=c('ensembl_transcript_id_version',
                         'external_gene_name'),
            mart = dog.anno)
Tx.dog <- as_tibble(Tx.dog)
#we need to rename the two columns we just retreived from biomart
Tx.dog <- dplyr::rename(Tx.dog, target_id = ensembl_transcript_id_version, 
                    gene_name = external_gene_name)
# ==============================================================================

# import Kallisto transcript counts into R using Tximport ----
Txi_gene <- tximport(path, 
                     type = "kallisto", 
                     tx2gene = Tx, 
                     txOut = FALSE, #How does the result change if this =FALSE vs =TRUE?
                     countsFromAbundance = "lengthScaledTPM",
                     ignoreTxVersion = TRUE)
beep(sound = 6)  # just for fun

#take a look at the type of object you just created
class(Txi_gene)
names(Txi_gene)

# if you want to write your counts or TPM to a file on your harddrive
# if you exported transcript level data and want to append your gene symbols to the data frame
Txi_trans <- as_tibble(Txi_gene$counts, rownames = "target_id")
Txi_trans <- left_join(Txi_trans, Tx)
write.csv(Txi_trans, "../Txi_trans1011.csv")
